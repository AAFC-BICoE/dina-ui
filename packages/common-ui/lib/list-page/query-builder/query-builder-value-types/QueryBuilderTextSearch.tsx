import React from "react";
import {
  includedTypeQuery,
  matchQuery,
  termQuery,
  existsQuery,
  prefixQuery,
  suffixQuery,
  infixQuery
} from "../query-builder-elastic-search/QueryBuilderElasticSearchExport";
import { TransformToDSLProps } from "../../types";
import { useIntl } from "react-intl";

interface QueryRowTextSearchProps {
  /**
   * Current match type being used.
   */
  matchType?: string;

  /**
   * Retrieve the current value from the Query Builder.
   */
  value?: string;

  /**
   * Pass the selected value to the Query Builder to store.
   */
  setValue?: (fieldPath: string) => void;
}

export default function QueryRowTextSearch({
  matchType,
  value,
  setValue
}: QueryRowTextSearchProps) {
  const { formatMessage } = useIntl();

  return (
    <>
      {/* Depending on the matchType, it changes the rest of the query row. */}
      {matchType !== "empty" && matchType !== "notEmpty" && (
        <input
          type="text"
          value={value ?? ""}
          onChange={(newValue) => setValue?.(newValue?.target?.value)}
          className="form-control"
          placeholder={formatMessage({
            id: "queryBuilder_value_text_placeholder"
          })}
        />
      )}
    </>
  );
}

/**
 * Using the query row for a text search, generate the elastic search request to be made.
 */
export function transformTextSearchToDSL({
  operation,
  value,
  fieldInfo,
  fieldPath
}: TransformToDSLProps): any {
  if (!fieldInfo) {
    return {};
  }

  const { distinctTerm, parentType, parentName } = fieldInfo;

  // Is the "Exact" option selected? (Or if auto suggestions are being used.)
  const isExactMatch: boolean = distinctTerm || operation === "exactMatch";

  switch (operation) {
    // Equals match type.
    case "equals":
    case "partialMatch":
    case "exactMatch":
      // Autocompletion expects to use the full text search.
      return parentType
        ? {
            nested: {
              path: "included",
              query: {
                bool: {
                  must: [
                    isExactMatch
                      ? termQuery(fieldPath, value, true)
                      : matchQuery(fieldPath, value),
                    includedTypeQuery(parentType)
                  ]
                }
              }
            }
          }
        : isExactMatch
        ? termQuery(fieldPath, value, true)
        : matchQuery(fieldPath, value);

    // Prefix partial match
    case "startsWith":
      return prefixQuery(fieldPath, value, parentType);

    // Infix partial match
    case "containsText":
      return infixQuery(fieldPath, value, parentType);

    // Suffix partial match
    case "endsWith":
      return suffixQuery(fieldPath, value, parentType);

    // Not equals match type.
    case "notEquals":
      return parentType
        ? {
            bool: {
              should: [
                // If the field does exist, then search for everything that does NOT match the term.
                {
                  nested: {
                    path: "included",
                    query: {
                      bool: {
                        must_not: isExactMatch
                          ? termQuery(fieldPath, value, true)
                          : matchQuery(fieldPath, value),
                        must: includedTypeQuery(parentType)
                      }
                    }
                  }
                },

                // If it's included but the field doesn't exist, then it's not equal either.
                {
                  nested: {
                    path: "included",
                    query: {
                      bool: {
                        must_not: existsQuery(fieldPath),
                        must: includedTypeQuery(parentType)
                      }
                    }
                  }
                },

                // And if it's not included, then it's not equal either.
                {
                  bool: {
                    must_not: existsQuery(
                      "data.relationships." + parentName + ".data.id"
                    )
                  }
                }
              ]
            }
          }
        : {
            bool: {
              should: [
                {
                  bool: {
                    must_not: isExactMatch
                      ? termQuery(fieldPath, value, true)
                      : matchQuery(fieldPath, value)
                  }
                },
                {
                  bool: {
                    must_not: existsQuery(fieldPath)
                  }
                }
              ]
            }
          };

    // Empty values only. (only if the value is not mandatory)
    case "empty":
      return parentType
        ? {
            bool: {
              should: [
                {
                  bool: {
                    should: [
                      {
                        bool: {
                          must_not: {
                            nested: {
                              path: "included",
                              query: {
                                bool: {
                                  must: [
                                    existsQuery(fieldPath),
                                    includedTypeQuery(parentType)
                                  ]
                                }
                              }
                            }
                          }
                        }
                      },
                      {
                        nested: {
                          path: "included",
                          query: {
                            bool: {
                              must: [
                                termQuery(fieldPath, "", true),
                                includedTypeQuery(parentType)
                              ]
                            }
                          }
                        }
                      }
                    ]
                  }
                },
                {
                  bool: {
                    must_not: existsQuery(
                      "data.relationships." + parentName + ".data.id"
                    )
                  }
                }
              ]
            }
          }
        : {
            bool: {
              should: [
                {
                  bool: {
                    must_not: existsQuery(fieldPath)
                  }
                },
                {
                  bool: {
                    must: termQuery(fieldPath, "", true)
                  }
                }
              ]
            }
          };

    // Not empty values only. (only if the value is not mandatory)
    case "notEmpty":
      return parentType
        ? {
            nested: {
              path: "included",
              query: {
                bool: {
                  must_not: termQuery(fieldPath, "", true),
                  must: [includedTypeQuery(parentType), existsQuery(fieldPath)]
                }
              }
            }
          }
        : {
            bool: {
              must: existsQuery(fieldPath),
              must_not: termQuery(fieldPath, "", true)
            }
          };

    // Unsupported case.
    default:
      throw new Error("Unsupported operation for the text widget.");
  }
}
