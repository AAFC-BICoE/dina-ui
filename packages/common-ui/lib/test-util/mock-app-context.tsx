import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { useMemo, useRef } from "react";
import { SWRConfig } from "swr";
import { PartialDeep } from "type-fest";
import { AccountContextI, AccountProvider } from "../account/AccountProvider";
import {
  ApiClientI,
  ApiClientImpl,
  ApiClientProvider
} from "../api-client/ApiClientContext";
import { CommonUIIntlProvider } from "../intl/common-ui-intl";
import { ModalProvider } from "../modal/modal";

interface MockAppContextProviderProps {
  apiContext?: PartialDeep<ApiClientI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext = { apiClient: { get: () => undefined as any } },
  children
}: MockAppContextProviderProps) {
  const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = useMemo(
    () => ({
      authenticated: true,
      groupNames: ["aafc", "cnc"],
      initialized: true,
      login: noop,
      logout: noop,
      roles: ["user"],
      token: "test-token",
      username: "test-user",
      isAdmin: false
    }),
    []
  );

  const DEFAULT_API_CONTEXT_VALUE = useMemo(
    () =>
      new ApiClientImpl({
        newId: () => "00000000-0000-0000-0000-000000000000"
      }),
    []
  );

  const apiContextWithWarnings = {
    ...apiContext,
    // Add a warning when bulkGet doesn't return anything in a test:
    bulkGet: async (paths: string[], options) => {
      const resources = await apiContext?.bulkGet?.(paths, options);
      if (!resources) {
        console.warn("No response returned for bulkGet paths: ", paths);
      } else {
        resources.forEach((resource, index) => {
          if (!resource) {
            console.warn("No value returned for bulkGet path: ", paths[index]);
          }
        });
      }
      return resources;
    }
  };

  const modalWrapperRef = useRef<HTMLDivElement>(null);

  return (
    <SWRConfig
      // Reset SWR cache between tests.
      value={{ provider: () => new Map() }}
    >
      <AccountProvider
        value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
      >
        <ApiClientProvider
          value={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContextWithWarnings)}
        >
          <CommonUIIntlProvider>
            <div ref={modalWrapperRef}>
              <ModalProvider appElement={modalWrapperRef.current}>
                {children}
              </ModalProvider>
            </div>
          </CommonUIIntlProvider>
        </ApiClientProvider>
      </AccountProvider>
    </SWRConfig>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
