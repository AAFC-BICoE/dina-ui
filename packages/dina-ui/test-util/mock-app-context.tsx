import {
  AccountContextI,
  AccountProvider,
  ApiClientI,
  ApiClientImpl,
  ApiClientProvider,
  ModalProvider
} from "common-ui";
import { mount } from "enzyme";
import { merge, noop } from "lodash";
import { useMemo, useRef } from "react";
import { DndProvider } from "react-dnd-cjs";
import HTML5Backend from "react-dnd-html5-backend-cjs";
import { SWRConfig } from "swr";
import { PartialDeep } from "type-fest";
import { FileUploadProviderImpl } from "../components/object-store/file-upload/FileUploadProvider";
import { DinaIntlProvider } from "../intl/dina-ui-intl";

interface MockAppContextProviderProps {
  apiContext?: PartialDeep<ApiClientI>;
  accountContext?: Partial<AccountContextI>;
  children?: React.ReactNode;
}

/**
 * Wraps a test-rendered component to provide the contexts that would be available in
 * the application.
 */
export function MockAppContextProvider({
  accountContext,
  apiContext = { apiClient: { get: () => undefined as any } },
  children
}: MockAppContextProviderProps) {
  const DEFAULT_MOCK_ACCOUNT_CONTEXT: AccountContextI = useMemo(
    () => ({
      authenticated: true,
      groupNames: ["aafc", "cnc"],
      initialized: true,
      login: noop,
      logout: noop,
      roles: ["user"],
      token: "test-token",
      username: "test-user",
      isAdmin: false
    }),
    []
  );

  const DEFAULT_API_CONTEXT_VALUE = useMemo(
    () =>
      new ApiClientImpl({
        newId: () => "00000000-0000-0000-0000-000000000000"
      }),
    []
  );

  const apiContextWithWarnings = {
    ...apiContext,
    // Add a warning when bulkGet doesn't return anything in a test:
    bulkGet: async (paths: string[], options) => {
      const resources = await apiContext?.bulkGet?.(paths, options);
      if (!resources) {
        console.warn("No response returned for bulkGet paths: ", paths);
      } else {
        resources.forEach((resource, index) => {
          if (!resource) {
            console.warn("No value returned for bulkGet path: ", paths[index]);
          }
        });
      }
      return resources;
    }
  };

  const modalWrapperRef = useRef<HTMLDivElement>(null);

  return (
    <SWRConfig
      // Reset SWR cache between tests.
      value={{ provider: () => new Map() }}
    >
      <AccountProvider
        value={{ ...DEFAULT_MOCK_ACCOUNT_CONTEXT, ...accountContext }}
      >
        <ApiClientProvider
          value={merge({}, DEFAULT_API_CONTEXT_VALUE, apiContextWithWarnings)}
        >
          <FileUploadProviderImpl>
            <DinaIntlProvider>
              <DndProvider backend={HTML5Backend}>
                <div ref={modalWrapperRef}>
                  <ModalProvider appElement={modalWrapperRef.current}>
                    {children}
                  </ModalProvider>
                </div>
              </DndProvider>
            </DinaIntlProvider>
          </FileUploadProviderImpl>
        </ApiClientProvider>
      </AccountProvider>
    </SWRConfig>
  );
}

/**
 * Helper function to get a test wrapper with the required context providers.
 */
export function mountWithAppContext(
  element: React.ReactNode,
  mockAppContextProviderProps?: MockAppContextProviderProps
) {
  return mount(
    <MockAppContextProvider {...mockAppContextProviderProps}>
      {element}
    </MockAppContextProvider>
  );
}
