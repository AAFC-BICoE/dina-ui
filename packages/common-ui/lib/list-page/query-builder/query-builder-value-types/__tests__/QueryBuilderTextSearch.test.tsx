import { mountWithAppContext } from "common-ui/lib/test-util/mock-app-context";
import QueryBuilderTextSearch, {
  transformTextSearchToDSL
} from "../QueryBuilderTextSearch";

describe("QueryBuilderTextSearch", () => {
  describe("QueryBuilderTextSearch Component", () => {
    it("Display field if match type is equals", async () => {
      // This test will just ensure the layout does not change unexpectedly.
      // Any changes to the layout, the snapshots will need to be updated.
      const textSearchEquals = mountWithAppContext(
        <QueryBuilderTextSearch
          matchType="equals"
          value="test"
          setValue={jest.fn}
        />
      );

      // Expect a snapshot with the text field being displayed.
      expect(
        textSearchEquals.find(QueryBuilderTextSearch).debug()
      ).toMatchSnapshot(
        "Expect text field to be displayed since match type is equals"
      );

      const textSearchEmpty = mountWithAppContext(
        <QueryBuilderTextSearch
          matchType="empty"
          value="test"
          setValue={jest.fn}
        />
      );

      // Expect a snapshot without the text field being displayed.
      expect(
        textSearchEmpty.find(QueryBuilderTextSearch).debug()
      ).toMatchSnapshot(
        "Expect text field not to be displayed since the match type is not equals"
      );
    });
  });

  describe("transformTextSearchToDSL function", () => {
    describe("Equals operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "equals",
            value: "text search",
            fieldInfo: {
              label: "name",
              value: "collection.name",
              type: "text",
              path: "attributes",
              parentName: "collection",
              parentType: "collection",
              parentPath: "included",
              distinctTerm: true
            } as any,
            fieldPath: "included.attributes.name",
            queryType: "equals"
          })
        ).toMatchSnapshot();
      });

      test("With relationship containing complex path as field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "exactMatch",
            value: "text search",
            fieldInfo: {
              label: "determination.scientificName",
              parentName: "organism",
              parentPath: "included",
              parentType: "organism",
              path: "attributes.determination",
              type: "text",
              value: "organism.determination.scientificName"
            } as any,
            fieldPath: "included.attributes.determination.scientificName",
            queryType: "exactMatch"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "equals",
            value: "text search",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.textField",
            queryType: "equals"
          })
        ).toMatchSnapshot();
      });
    });

    describe("startsWith operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "startsWith",
            value: "text search",
            fieldInfo: {
              label: "name",
              value: "collection.name",
              type: "text",
              path: "attributes",
              parentName: "collection",
              parentType: "collection",
              parentPath: "included",
              distinctTerm: true
            } as any,
            fieldPath: "included.attributes.name",
            queryType: "startsWith"
          })
        ).toMatchSnapshot();
      });

      test("With relationship containing complex path as field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "startsWith",
            value: "text",
            fieldInfo: {
              label: "determination.scientificName",
              parentName: "organism",
              parentPath: "included",
              parentType: "organism",
              path: "attributes.determination",
              type: "text",
              value: "organism.determination.scientificName"
            } as any,
            fieldPath: "included.attributes.determination.scientificName",
            queryType: "startsWith"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "startsWith",
            value: "text search",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.textField",
            queryType: "startsWith"
          })
        ).toMatchSnapshot();
      });
    });

    describe("ContainsText (Infix) operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "containsText",
            value: "text search",
            fieldInfo: {
              label: "name",
              value: "collection.name",
              type: "text",
              path: "attributes",
              parentName: "collection",
              parentType: "collection",
              parentPath: "included",
              distinctTerm: true
            } as any,
            fieldPath: "included.attributes.name",
            queryType: "containsText"
          })
        ).toMatchSnapshot();
      });

      test("With relationship containing complex path as field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "containsText",
            value: "text",
            fieldInfo: {
              label: "determination.scientificName",
              parentName: "organism",
              parentPath: "included",
              parentType: "organism",
              path: "attributes.determination",
              type: "text",
              value: "organism.determination.scientificName"
            } as any,
            fieldPath: "included.attributes.determination.scientificName",
            queryType: "containsText"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "containsText",
            value: "text search",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.textField",
            queryType: "containsText"
          })
        ).toMatchSnapshot();
      });
    });

    describe("EndsWith operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "endsWith",
            value: "text search",
            fieldInfo: {
              label: "name",
              value: "collection.name",
              type: "text",
              path: "attributes",
              parentName: "collection",
              parentType: "collection",
              parentPath: "included",
              distinctTerm: true
            } as any,
            fieldPath: "included.attributes.name",
            queryType: "endsWith"
          })
        ).toMatchSnapshot();
      });

      test("With relationship containing complex path as field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "endsWith",
            value: "text",
            fieldInfo: {
              label: "determination.scientificName",
              parentName: "organism",
              parentPath: "included",
              parentType: "organism",
              path: "attributes.determination",
              type: "text",
              value: "organism.determination.scientificName"
            } as any,
            fieldPath: "included.attributes.determination.scientificName",
            queryType: "endsWith"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "endsWith",
            value: "text search",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.textField",
            queryType: "endsWith"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Partial Match operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "partialMatch",
            value: "text search",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "partialMatch"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "partialMatch",
            value: "text search",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.textField",
            queryType: "partialMatch"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Exact Match operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "exactMatch",
            value: "text search",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "exactMatch"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "exactMatch",
            value: "text search",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.textField",
            queryType: "exactMatch"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Not Equals operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "notEquals",
            value: "text search",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "notEquals"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "notEquals",
            value: "text search",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.textField",
            queryType: "notEquals"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Empty operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "empty",
            value: "text search",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "empty"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "empty",
            value: "text search",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.textField",
            queryType: "empty"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Not empty operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "notEmpty",
            value: "text search",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "notEmpty"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "notEmpty",
            value: "text search",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.textField",
            queryType: "notEmpty"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Equals operation", () => {
      test("With relationship as field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "equals",
            value: "false",
            fieldInfo: {
              parentType: "collection",
              parentName: "collection"
            } as any,
            fieldPath: "includes.name",
            queryType: "equals"
          })
        ).toMatchSnapshot();
      });

      test("Normal field", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "equals",
            value: "true",
            fieldInfo: {} as any,
            fieldPath: "data.attributes.textField",
            queryType: "equals"
          })
        ).toMatchSnapshot();
      });
    });

    describe("Edge cases", () => {
      test("If no field value is provided, nothing should be generated.", async () => {
        expect(
          transformTextSearchToDSL({
            operation: "exactMatch",
            value: "false",
            fieldInfo: undefined,
            fieldPath: "includes.name",
            queryType: "exactMatch"
          })
        ).toMatchSnapshot();
      });
    });
  });
});
