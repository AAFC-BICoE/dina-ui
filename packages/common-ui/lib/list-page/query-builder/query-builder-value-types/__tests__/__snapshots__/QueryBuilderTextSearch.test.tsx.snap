// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryBuilderTextSearch QueryBuilderTextSearch Component Display field if match type is equals: Expect text field not to be displayed since the match type is not equals 1`] = `"<QueryRowTextSearch matchType=\\"empty\\" value=\\"test\\" setValue={[Function: bound fn]} />"`;

exports[`QueryBuilderTextSearch QueryBuilderTextSearch Component Display field if match type is equals: Expect text field to be displayed since match type is equals 1`] = `
"<QueryRowTextSearch matchType=\\"equals\\" value=\\"test\\" setValue={[Function: bound fn]}>
  <input type=\\"text\\" value=\\"test\\" onChange={[Function: onChange]} className=\\"form-control\\" placeholder=\\"queryBuilder_value_text_placeholder\\" />
</QueryRowTextSearch>"
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function ContainsText (Infix) operation Normal field 1`] = `
Object {
  "match": Object {
    "data.attributes.textField.infix": Object {
      "query": "text search",
    },
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function ContainsText (Infix) operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "match": Object {
              "included.attributes.name.infix": Object {
                "query": "text search",
              },
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function ContainsText (Infix) operation With relationship containing complex path as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "match": Object {
              "included.attributes.determination.scientificName.infix": Object {
                "query": "text",
              },
            },
          },
          Object {
            "term": Object {
              "included.type": "organism",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Edge cases If no field value is provided, nothing should be generated. 1`] = `Object {}`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Empty operation Normal field 1`] = `
Object {
  "bool": Object {
    "should": Array [
      Object {
        "bool": Object {
          "must_not": Object {
            "exists": Object {
              "field": "data.attributes.textField",
            },
          },
        },
      },
      Object {
        "bool": Object {
          "must": Object {
            "term": Object {
              "data.attributes.textField.keyword": "",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Empty operation With relationship as field 1`] = `
Object {
  "bool": Object {
    "should": Array [
      Object {
        "bool": Object {
          "should": Array [
            Object {
              "bool": Object {
                "must_not": Object {
                  "nested": Object {
                    "path": "included",
                    "query": Object {
                      "bool": Object {
                        "must": Array [
                          Object {
                            "exists": Object {
                              "field": "includes.name",
                            },
                          },
                          Object {
                            "term": Object {
                              "included.type": "collection",
                            },
                          },
                        ],
                      },
                    },
                  },
                },
              },
            },
            Object {
              "nested": Object {
                "path": "included",
                "query": Object {
                  "bool": Object {
                    "must": Array [
                      Object {
                        "term": Object {
                          "includes.name.keyword": "",
                        },
                      },
                      Object {
                        "term": Object {
                          "included.type": "collection",
                        },
                      },
                    ],
                  },
                },
              },
            },
          ],
        },
      },
      Object {
        "bool": Object {
          "must_not": Object {
            "exists": Object {
              "field": "data.relationships.collection.data.id",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function EndsWith operation Normal field 1`] = `
Object {
  "prefix": Object {
    "data.attributes.textField.prefix_reverse": "hcraes txet",
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function EndsWith operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "prefix": Object {
              "included.attributes.name.prefix_reverse": "hcraes txet",
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function EndsWith operation With relationship containing complex path as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "prefix": Object {
              "included.attributes.determination.scientificName.prefix_reverse": "txet",
            },
          },
          Object {
            "term": Object {
              "included.type": "organism",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Equals operation Normal field 1`] = `
Object {
  "match": Object {
    "data.attributes.textField": "text search",
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Equals operation Normal field 2`] = `
Object {
  "match": Object {
    "data.attributes.textField": "true",
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Equals operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "term": Object {
              "included.attributes.name.keyword": "text search",
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Equals operation With relationship as field 2`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "match": Object {
              "includes.name": "false",
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Equals operation With relationship containing complex path as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "term": Object {
              "included.attributes.determination.scientificName.keyword": "text search",
            },
          },
          Object {
            "term": Object {
              "included.type": "organism",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Exact Match operation Normal field 1`] = `
Object {
  "term": Object {
    "data.attributes.textField.keyword": "text search",
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Exact Match operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "term": Object {
              "includes.name.keyword": "text search",
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Not Equals operation Normal field 1`] = `
Object {
  "bool": Object {
    "should": Array [
      Object {
        "bool": Object {
          "must_not": Object {
            "match": Object {
              "data.attributes.textField": "text search",
            },
          },
        },
      },
      Object {
        "bool": Object {
          "must_not": Object {
            "exists": Object {
              "field": "data.attributes.textField",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Not Equals operation With relationship as field 1`] = `
Object {
  "bool": Object {
    "should": Array [
      Object {
        "nested": Object {
          "path": "included",
          "query": Object {
            "bool": Object {
              "must": Object {
                "term": Object {
                  "included.type": "collection",
                },
              },
              "must_not": Object {
                "match": Object {
                  "includes.name": "text search",
                },
              },
            },
          },
        },
      },
      Object {
        "nested": Object {
          "path": "included",
          "query": Object {
            "bool": Object {
              "must": Object {
                "term": Object {
                  "included.type": "collection",
                },
              },
              "must_not": Object {
                "exists": Object {
                  "field": "includes.name",
                },
              },
            },
          },
        },
      },
      Object {
        "bool": Object {
          "must_not": Object {
            "exists": Object {
              "field": "data.relationships.collection.data.id",
            },
          },
        },
      },
    ],
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Not empty operation Normal field 1`] = `
Object {
  "bool": Object {
    "must": Object {
      "exists": Object {
        "field": "data.attributes.textField",
      },
    },
    "must_not": Object {
      "term": Object {
        "data.attributes.textField.keyword": "",
      },
    },
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Not empty operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
          Object {
            "exists": Object {
              "field": "includes.name",
            },
          },
        ],
        "must_not": Object {
          "term": Object {
            "includes.name.keyword": "",
          },
        },
      },
    },
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Partial Match operation Normal field 1`] = `
Object {
  "match": Object {
    "data.attributes.textField": "text search",
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function Partial Match operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "match": Object {
              "includes.name": "text search",
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function startsWith operation Normal field 1`] = `
Object {
  "prefix": Object {
    "data.attributes.textField.prefix": "text search",
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function startsWith operation With relationship as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "prefix": Object {
              "included.attributes.name.prefix": "text search",
            },
          },
          Object {
            "term": Object {
              "included.type": "collection",
            },
          },
        ],
      },
    },
  },
}
`;

exports[`QueryBuilderTextSearch transformTextSearchToDSL function startsWith operation With relationship containing complex path as field 1`] = `
Object {
  "nested": Object {
    "path": "included",
    "query": Object {
      "bool": Object {
        "must": Array [
          Object {
            "prefix": Object {
              "included.attributes.determination.scientificName.prefix": "text",
            },
          },
          Object {
            "term": Object {
              "included.type": "organism",
            },
          },
        ],
      },
    },
  },
}
`;
